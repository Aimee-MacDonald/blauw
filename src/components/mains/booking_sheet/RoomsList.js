import React from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {getRoomGroupings} from '../../../state_management/selectors/rooms'

export const RoomsList = ({groups, groupFlags, setGroupFlags}) => (
  <StyledRoomsList>
    {groups.map(({groupName, rooms}) => (
      <StyledGroup key={groupName}>
        <h3 onClick={e => setGroupFlags({
          ...groupFlags,
          [groupName]: {
            ...groupFlags[groupName],
            open: !groupFlags[groupName].open
          }
        })}>{groupName}</h3>

        {groupFlags[groupName].open && rooms.map(({roomName, maxPax}) => {
          const beds = []
          for(let i = 0; i < maxPax; i++) beds.push(`Bed ${i + 1}`)

          return(
            <StyledRooms key={roomName}>
              <h3 onClick={e => setGroupFlags({
                ...groupFlags,
                [groupName]: {        
                  ...groupFlags[groupName],
                  rooms: {
                    ...groupFlags[groupName].rooms,
                    [roomName]: !groupFlags[groupName].rooms[roomName]
                  }
                }
              })}>{roomName}</h3>

              {groupFlags[groupName].rooms[roomName] && beds.map(bed => (
                <StyledBed key={`${roomName}${bed}`}>
                  <h3>{bed}</h3>
                </StyledBed>
              ))}
            </StyledRooms>
          )
        })}
      </StyledGroup>
    ))}
  </StyledRoomsList>
)

const StyledRoomsList = styled.div`
  background-color: lime;
  height: 100%;
  overflow-y: hidden;
`

const StyledGroup = styled.div`
  background-color: orangered;

  h3{
    height: 32px;
    display: grid;
    place-items: center center;
    cursor: pointer;
  }
`

const StyledRooms = styled.div`
  background-color: magenta;
`

const StyledBed = styled.div`
  background-color: violet;

  h3{
    place-items: center right;
    padding-right: 1rem;
  }
`

const mapStateToProps = ({rooms}) => ({groups: getRoomGroupings(rooms)})

export default connect(mapStateToProps)(RoomsList)