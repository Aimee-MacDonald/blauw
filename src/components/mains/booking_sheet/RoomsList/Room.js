import React from 'react'
import styled from 'styled-components'

export const Room = props => {
  const beds = []

  if(props.room.showBeds){
    for(let i = 0; i < props.room.maxPax; i++){
      beds.push(`Bed ${i + 1}`)
    }
  }

  return(
    <StyledRoom {...props}>
      <h3 onClick={props.toggleRoom}>{props.room.name}</h3>

      {props.room.showBeds && props.roomFlag &&
        beds.map((bed, i) => (
          <StyledBed key={`${props.room._id}_${i}`}>
            <h3>{bed}</h3>
          </StyledBed>
        ))
      }
    </StyledRoom>
  )
}

const StyledRoom = styled.div`
  background-color: red;
`

const StyledBed = styled.div`
  background-color: green;
`

export default Room