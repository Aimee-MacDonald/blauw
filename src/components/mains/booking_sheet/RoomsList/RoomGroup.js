import React from 'react'
import styled from 'styled-components'

import Room from './Room'

export const RoomGroup = props => (
  <StyledRoomGroup {...props}>
    {props.group && props.group.name !== 'ungrouped' &&
      <h3 onClick={props.toggleGroup}>{props.group.name}</h3>
    }

    {props.groupFlags && props.groupFlags.open && props.group.rooms.map(room => (
      <Room
        key={room._id}
        room={room}
        roomFlag={props.groupFlags.rooms[room.name]}
        toggleRoom={() => props.toggleRoom(room.name)}
        hoveredRoom={props.hoveredRoom}
      />
    ))}
  </StyledRoomGroup>
)

const StyledRoomGroup = styled.div`
  background-color: ${props => props.group.name === props.hoveredRoom ? 'cornflowerblue' : 'hotpink'};

  h3{
    height: 32px;
    display: grid;
    place-items: center center;
  }
`

export default RoomGroup