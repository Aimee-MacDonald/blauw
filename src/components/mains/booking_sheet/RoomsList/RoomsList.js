import React, {useEffect} from 'react'
import styled from 'styled-components'
import {connect} from 'react-redux'

import {getDefaultRoomFlags, getRoomGroupings, getHoveredRoom} from '../../../../state_management/selectors/rooms'

import RoomGroup from './RoomGroup'

export const RoomsList = ({groups, groupFlags, setGroupFlags, hoveredCell}) => {
  useEffect(() => setGroupFlags(getDefaultRoomFlags(groups)), [groups])

  const toggleGroup = groupName => {
    setGroupFlags({
      ...groupFlags,
      [groupName]: {
        ...groupFlags[groupName],
        open: !groupFlags[groupName].open
      }
    })
  }

  const toggleRoom = (groupName, roomName) => {
    setGroupFlags({
      ...groupFlags,
      [groupName]: {
        ...groupFlags[groupName],
        rooms: {
          ...groupFlags[groupName].rooms,
          [roomName]: !groupFlags[groupName].rooms[roomName]
        }
      }
    })
  }

  return(
    <StyledRoomsList>
      {groups.map(group => (
        <RoomGroup
          key={group.name}
          group={group}
          groupFlags={groupFlags[group.name]}
          toggleGroup={() => toggleGroup(group.name)}
          toggleRoom={roomName => toggleRoom(group.name, roomName)}
          hoveredRoom={getHoveredRoom(groups, groupFlags, hoveredCell)}
        />
      ))}
    </StyledRoomsList>
  )
}

const StyledRoomsList = styled.div`
  background-color: lime;
  height: 100%;
  overflow-y: hidden;
`

const mapStateToProps = ({rooms}) => ({groups: getRoomGroupings(rooms)})

export default connect(mapStateToProps)(RoomsList)