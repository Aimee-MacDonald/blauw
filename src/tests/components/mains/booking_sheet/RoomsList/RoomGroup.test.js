import React from 'react'
import {shallow} from 'enzyme'

import {RoomGroup} from '../../../../../components/mains/booking_sheet/RoomsList/RoomGroup'

test('Render a group of rooms', () => {
  const wrapper = shallow(<RoomGroup group={{name: 'abcd1234'}}/>)
  expect(wrapper.getElement()).toMatchSnapshot()
})