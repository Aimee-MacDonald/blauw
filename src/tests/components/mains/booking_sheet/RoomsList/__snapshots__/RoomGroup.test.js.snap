// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a group of rooms 1`] = `
<styled.div
  group={
    Object {
      "name": "abcd1234",
    }
  }
>
  <h3>
    abcd1234
  </h3>
</styled.div>
`;
